presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/InMemoryHashAggregationBuilder.java

```diff
@@ -348,14 +395,38 @@ public Type getType()
 
         public void processPage(GroupByIdBlock groupIds, Page page)
         {
-            if (step.isInputRaw()) {
+            if (!this.orderByChannels.isEmpty()) {
+                checkArgument(step == Step.SINGLE && pageIndex.isPresent(), "ORDER BY in aggregations is supported only for single step aggregations");
+                pageIndex.ifPresent(value -> value.addPage(page));
+            }
+            else if (step.isInputRaw()) {
                 aggregation.addInput(groupIds, page);
             }
             else {
                 aggregation.addIntermediate(groupIds, page.getBlock(intermediateChannel));
             }
         }
 
+        public void processOrderBy()
+        {
+            if (this.pageIndex.isPresent()) {
+                checkArgument(groupByHash.isPresent(), "No groupByHash");
+                PagesIndex pageIndex = this.pageIndex.get();
+                pageIndex.sort(orderByChannels, orderings);
+                int currentPosition = 0;
+                while (true) {
+                    pageBuilder.reset();
+                    currentPosition = pageIndex.buildPage(currentPosition, orderByOutputChannels, pageBuilder);
```

Comment:
Ah.. i didn't notice you're not pulling all channels (nor I understand why). You'd need to enhance https://github.com/prestodb/presto/commit/9e2bffcda644d7a21547d1e6c9ba0e9288474ead with ability to specify the channels. Still it looks like writing this once (even if parameterized) & reusing would be nice.